modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java

```diff
@@ -9,145 +9,153 @@
 package org.elasticsearch.repositories.gcs;
 
 import com.google.api.client.http.HttpResponse;
-import com.google.cloud.BaseService;
-import com.google.cloud.RetryHelper;
 import com.google.cloud.storage.BlobId;
 import com.google.cloud.storage.StorageException;
+import com.google.cloud.storage.StorageRetryStrategy;
 
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.elasticsearch.common.blobstore.OperationPurpose;
-import org.elasticsearch.core.IOUtils;
+import org.elasticsearch.common.blobstore.RetryingInputStream;
+import org.elasticsearch.core.Nullable;
 import org.elasticsearch.repositories.blobstore.RequestedRangeNotSatisfiedException;
 import org.elasticsearch.rest.RestStatus;
 
 import java.io.FilterInputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.nio.file.NoSuchFileException;
-import java.util.ArrayList;
-import java.util.List;
-
-import static org.elasticsearch.core.Strings.format;
 
 /**
  * Wrapper around reads from GCS that will retry blob downloads that fail part-way through, resuming from where the failure occurred.
  * This should be handled by the SDK but it isn't today. This should be revisited in the future (e.g. before removing
  * the {@link org.elasticsearch.Version#V_7_0_0} version constant) and removed if the SDK handles retries itself in the future.
  */
-class GoogleCloudStorageRetryingInputStream extends InputStream {
+class GoogleCloudStorageRetryingInputStream extends RetryingInputStream<Long> {
 
     private static final Logger logger = LogManager.getLogger(GoogleCloudStorageRetryingInputStream.class);
-
-    static final int MAX_SUPPRESSED_EXCEPTIONS = 10;
-
-    private final OperationPurpose purpose;
-    private final MeteredStorage client;
-    private final BlobId blobId;
-    private final long start;
-    private final long end;
-    private final int maxAttempts;
-    private InputStream currentStream;
-    private int attempt = 1;
-    private List<StorageException> failures = new ArrayList<>(MAX_SUPPRESSED_EXCEPTIONS);
-    private long currentOffset;
-    private boolean closed;
-    private Long lastGeneration;
+    private static final StorageRetryStrategy STORAGE_RETRY_STRATEGY = GoogleCloudStorageService.createStorageRetryStrategy();
```

Comment:
The one we use is stateless, we might need to re-think this lifecycle if we switch to one that is not. You can't get it out of the client or `StorageOptions` as far as I could see.